gooddata-pandas/gooddata_pandas/result_convertor.py

```diff
@@ -0,0 +1,309 @@
+                # all the columns.
+                #
+                # page 'to the right' to get data from all columns, extend existing rows with that data
+                offset = [offset[0], result.next_page_start(dim=1)]
+                while True:
+                    result = response.read_result(offset=offset, limit=limit)
+                    acc.extend_existing_row_data(from_result=result)
+
+                    if load_headers_and_totals:
+                        acc.accumulate_headers(from_result=result, from_dim=1)
+                        acc.accumulate_grand_totals(from_result=result, paging_dim=1, response=response)
+
+                    if result.is_complete(dim=1):
+                        break
+
+                    offset = [offset[0], result.next_page_start(dim=1)]
+
+        if result.is_complete(dim=0):
+            break
+
+        offset = [result.next_page_start(dim=0), 0] if num_dims > 1 else [result.next_page_start(dim=0)]
+
+    return acc.result()
+
+
+def _create_header_mapper(
+    response: BareExecutionResponse, dim: int, label_overrides: LabelOverrides = {}
+) -> Callable[[int, Any], str]:
+    """
+    Prepares header mapper function which is doing header structures translations into appropriate label used
+    in a dataframe
+    :param response: Response structure to gather dimension header details
+    :param dim: dimension id
+    :param label_overrides: label overrides
+    :return: Mapper function
+    """
+    dim_descriptor = response.dimensions[dim]
+    attribute_labels = label_overrides.get("labels", {})
+    measure_labels = label_overrides.get("metrics", {})
+
+    def _mapper(header_idx: int, header: Any) -> str:
+        label = ""
+        if header is None:
+            pass
+        elif "attributeHeader" in header:
+            if "labelValue" in header["attributeHeader"]:
+                label = header["attributeHeader"]["labelValue"]
+            elif "labelName" in header["attributeHeader"]:
+                attr_local_id = header["attributeHeader"]["localIdentifier"]
+                if attr_local_id in attribute_labels:
+                    label = attribute_labels[attr_local_id]["title"]
+                else:
+                    label = header["attributeHeader"]["labelName"]
+        elif "measureHeader" in header:
+            measure_idx = header["measureHeader"]["measureIndex"]
+            measure_descriptor = dim_descriptor["headers"][header_idx]["measureGroupHeaders"][measure_idx]
+
+            if measure_descriptor["localIdentifier"] in measure_labels:
+                label = measure_labels[measure_descriptor["localIdentifier"]]["title"]
+            elif "name" in measure_descriptor:
+                label = measure_descriptor["name"]
+            else:
+                label = measure_descriptor["localIdentifier"]
+        elif "totalHeader" in header:
+            label = header["totalHeader"]["function"]
+        return label
+
+    return _mapper
```

Comment:
This function seems crazy to me. Is there a way to simplify it?

I do not fully understand why is not parameter header_idx optional in the case of this branch
```
        elif "attributeHeader" in header:
            if "labelValue" in header["attributeHeader"]:
                label = header["attributeHeader"]["labelValue"]
            elif "labelName" in header["attributeHeader"]:
                attr_local_id = header["attributeHeader"]["localIdentifier"]
                if attr_local_id in attribute_labels:
                    label = attribute_labels[attr_local_id]["title"]
                else:
                    label = header["attributeHeader"]["labelName"]
```

it is not needed at all.

Reply:
switched args order and make the index optional - however 2nd arg has to be always given (as None) anyway - seems like a limitation of mypy recognition